Tutorial 4: Characterising Benthic Biozones (#39)
r-rse/emodnet-bio-r-geo-tutorials: tutorials/tutorial-04.qmd

---
title: "4: Characterising Benthic Biozones"
subtitle: "Combining Habitat Maps, Bathymetry, Temperature, Species Occurrences and Body Size Data in the Gulf of Lion"
difficulty: "Advanced"
time: "~90 minutes"
data-type: "WFS + WCS + External sources"
description: "Learn to integrate EMODnet WFS and WCS services with external data sources including the Copernicus Marine Service and trait databases. Characterise benthic biozones in the Gulf of Lion using habitat composition, substrate type, environmental conditions, and body size distributions."
datasets: "Seabed habitat polygons, benthic species occurrences, bathymetry, bottom temperature, body size data"
df-print: paged
bibliography: references.bib
format:
  html:
    embed-resources: false
    code-link: true
---

## Introduction

> The Gulf of Lion, located in the northwestern Mediterranean off the coast of southern France, hosts a rich diversity of benthic communities shaped by the Rhône River delta, diverse substrates from sandy beaches to deep canyons, and strong environmental gradients. Characterising the biozones of this region, from the sunlit infralittoral to the deep bathyal, requires bringing together habitat maps, environmental data, species occurrences, and functional traits such as body size.

Author: The Wikipedia link should give enough context. Adding a locator map would be a new pattern across all tutorials, so I'll skip it for now.

-   Link species records to body size data from external databases
-   Use spatial joins to assign habitat and biozone attributes to species occurrences based on location
-   Join datasets using shared identifiers (AphiaIDs)
-   Extract zonal statistics from rasters using habitat polygons
-   Characterise benthic biozones using habitat, environmental, and trait data
-   Create interactive maps with the `leaflet` package

### Data Sources

**EMODnet WFS (Vector Data):**

| Layer | Service | Description |
|------------------|----------------------|--------------------------------|
| `eusm2025_eunis2019_full` | seabed_habitats_general_datasets_and_products | EU Seabed Map with EUNIS 2019 classification |
| `eb185_benthos_full_matched` | biology | Benthic species occurrence records from European seas |

**EMODnet WCS (Raster Data):**

| Coverage             | Service    | Description                   |
|----------------------|------------|-------------------------------|
| `emodnet__mean_2022` | bathymetry | Mean water depth (bathymetry) |

**External Sources:**

| Source | Description |
|----------------------------|--------------------------------------------|

> | [Copernicus Marine Service](https://marine.copernicus.eu/) | Bottom temperature from [MEDSEA_MULTIYEAR_PHY_006_004](https://data.marine.copernicus.eu/product/MEDSEA_MULTIYEAR_PHY_006_004/description) reanalysis |

Reviewer: Sorry for the late contribution, but now I'm reading this, it is probably safer/more stable to link to the DOI:
https://doi.org/10.48670/mds-00375

```

## Retrieving Seabed Habitat Data

Our first data source is seabed habitat polygons from EMODnet Seabed Habitats. These polygons define the spatial framework for our biozone characterisation, providing EUNIS habitat codes, substrate types, and biozone classifications across the study area.

### Connecting to EMODnet Seabed Habitats WFS

For an introduction to working with WFS services and the `emodnet.wfs` package, see [Tutorial 1](tutorial-01.qmd). Here we apply the same workflow to the Seabed Habitats service.

```{r}
#| label: init-habitats-wfs
habitat_wfs <- emodnet_init_wfs_client(
  service = "seabed_habitats_general_datasets_and_products"
)
```

### Exploring Habitat Layers

We need habitat polygons from the [EUSeaMap](https://emodnet.ec.europa.eu/en/seabed-habitats), the broad-scale predictive habitat map for European seas. The service contains many layers, so we filter by title to find the EUSeaMap products:

```{r}
#| label: explore-habitat-layers
#| cache: true
habitat_layers <- emodnet_get_wfs_info(habitat_wfs)

# Search for EUSeaMap layers by title
habitat_layers |>
  filter(grepl("EUSeaMap", title)) |>
  select(layer_name, title)
```

> The results show separate layers for energy, substrate, biozone, and two habitat classification systems (EUNIS 2007 and EUNIS 2019). We choose `eusm2025_eunis2019_full` because EUNIS 2019 is the current classification standard and this layer includes both the habitat code and the substrate and biozone attributes we need.

Reviewer: ```suggestion
The results show separate layers for energy, substrate, biozone, and two habitat classification systems (EUNIS 2007 and EUNIS 2019). We choose `eusm2025_eunis2019_full` because EUNIS 2019 is the more recent classification standard and this layer includes both the habitat code and the substrate and biozone attributes we need.
```

Author: Done.

### Understanding Habitat Classifications

Let's examine the attribute structure of the habitat layer:

```{r}
#| label: inspect-habitat-attributes
#| cache: true
habitat_attrs <- layer_attribute_descriptions(
  wfs = habitat_wfs,
  layer = "eusm2025_eunis2019_full"
)

habitat_attrs
```

Key attributes include:

> -   **`geom`**: Geometry column. WFS layers can use different names for this column, so we extract it programmatically below to construct CQL bounding box filters

Author: Added a note that `geom` is the only row with `geometry = TRUE`.

    colour = NA,
    alpha = 0.8
  ) +
  coord_sf(
    xlim = c(
      study_bbox_3035["xmin"],
      study_bbox_3035["xmax"]
    ),
    ylim = c(
      study_bbox_3035["ymin"],
      study_bbox_3035["ymax"]
    )
  ) +
  scale_fill_viridis_d(option = "turbo", name = NULL) +
  labs(title = "Seabed Habitats (EUNIS 2019)") +
  theme_minimal() +
  theme(
    legend.position = "bottom",
    legend.text = element_text(size = 9),
    legend.key.size = unit(0.4, "cm"),
    plot.margin = margin(2, 2, 2, 2)
  ) +
  guides(fill = guide_legend(ncol = 2, override.aes = list(alpha = 1)))
```

## Retrieving Species Occurrence Data

Next, we retrieve benthic species occurrence records from EMODnet Biology. These are geolocated records whose species names have also been matched to the World Register of Marine Species (WoRMS) via an AphiaID code, ensuring taxonomic consistency. Later, we will use a spatial join to link these point occurrences to the habitat polygons we just downloaded, assigning each record to a habitat type, biozone and substrate based on where it falls on the seabed map.

::: {.callout-tip collapse="true"}
## Taxonomic matching with the `worrms` package

> The data we use here are already matched to WoRMS, but if you need to match your own species lists to the WoRMS taxonomic backbone, the [`worrms`](https://cran.r-project.org/package=worrms) R package provides a programmatic interface to WoRMS. It supports fuzzy name matching, retrieval of accepted names and synonyms, and lookup of taxonomic hierarchies, useful for standardising species names across datasets.

Author: Done, switched to the pkgdown site link.

| Layer (e.g., `"eb185_benthos_full_matched"`) | **Layer** (a specific temporal resolution or subset) | `cmems_mod_med_phy-temp_my_4.2km_P1M-m` |
| Attributes (WFS) or Bands (WCS) | **Variables** (physical quantities stored in the array) | `bottomT`, `thetao` |
| 2D features or grids | **Multi-dimensional arrays** with longitude, latitude, depth, and time | 4D datacube |

Because the data is multi-dimensional, `cms_download_subset()` returns a [`stars`](../concepts.qmd#stars---spatiotemporal-arrays) object rather than the `sf` or `terra` objects returned by EMODnet services. For more background, see the [stars](../concepts.qmd#stars---spatiotemporal-arrays) and [zarr](../concepts.qmd#zarr-.zarr) sections in the Geospatial Concepts page.

The metadata returned by `CopernicusMarine` functions follows the STAC specification. For a general introduction to what STAC is and how catalogues, collections, and items relate, see the [STAC intro tutorial](https://stacspec.org/en/tutorials/intro-to-stac/). In practice, the deeply nested metadata structures can be challenging to navigate, but contain rich information about layers, variables, dimensions, and time ranges, as we explore below.

> The discovery workflow follows a similar pattern to EMODnet: list what is available, inspect metadata, then download a subset.

Author: Done, added links to both vignettes.

### Choosing a Data Product

Bottom temperature is typically available in physical reanalysis products, identified by `MULTIYEAR_PHY` in their product IDs. `cms_products_list()` lists all available products; we filter to physical reanalysis:

```{r}
#| label: explore-copernicus-products
#| cache: true
# List physical reanalysis products
cms_products_list() |>
  filter(grepl("MULTIYEAR_PHY", product_id)) |>
  select(product_id, title) |>
  knitr::kable()
```

A Copernicus **product** plays the same role as an EMODnet **service** (like `"biology"` or `"bathymetry"`): it is a thematic collection that groups related layers together. For small study areas, regional products typically offer higher resolution and are specifically validated for that region. For our Gulf of Lion study, the Mediterranean product ([MEDSEA_MULTIYEAR_PHY_006_004](https://data.marine.copernicus.eu/product/MEDSEA_MULTIYEAR_PHY_006_004/description)) provides \~4.2 km resolution compared to \~8 km for the global product.

### Browsing Available Variables

> Unlike OGC services where each layer contains a single dataset, Copernicus products are built on a datacube architecture. A single product can bundle multiple physical variables (temperature, salinity, currents) across multiple layers. We can get a high-level overview of all variables available across the entire product using `cms_product_details()`:

Reviewer: ```suggestion
Unlike OGC services where each layer contains a single dataset, Copernicus products are built on a datacube architecture. A single product can bundle multiple physical variables (temperature, salinity, currents) across multiple layers, where each layer can have distinct dimensions. We can get a high-level overview of all variables available across the entire product using `cms_product_details()`:
```

```{r}
#| label: inspect-cube-variables
# Examine the metadata structure of the first variable in the first layer
str(med_services$properties[[1]]$`cube:variables`[[1]], max.level = 2)
```

The standard STAC datacube fields include `dimensions` (which axes the variable spans), `type` (`"data"` for measured values), and `unit`. In standard STAC, each variable is identified by its key name in the `cube:variables` map. Copernicus Marine also stores this as an explicit `id` field inside each variable object, along with other Copernicus-specific additions: `name$en` (a human-readable description), `standardName` (the [CF convention](https://cfconventions.org/Data/cf-standard-names/current/build/cf-standard-name-table.html) name), and visualisation hints like `colormapId` and value ranges.

> We can use these fields to build a clear mapping of layers to their variables. Since we need bottom temperature, we filter to the temperature layers:

Reviewer: This code-chunk might be useful for implementing a more user-friendly presentation of a selection of meta-data.

```{r}
#| label: check-temp-layer-variables
# Extract variable details from each temperature layer
med_services |>
  filter(grepl("tem", id, ignore.case = TRUE)) |>
  rowwise() |>
  reframe(
    layer = id,
    variable = purrr::map_chr(properties$`cube:variables`, "id"),
    description = purrr::map_chr(properties$`cube:variables`, \(x) x$name$en),
    unit = purrr::map_chr(properties$`cube:variables`, "unit")
  ) |>
  knitr::kable()
```

> All temperature layers contain `bottomT` (sea floor potential temperature) and `thetao` (potential temperature through the water column). For comprehensive documentation of all variable names, layer naming conventions, and production methodology, see the [Product User Manual (PUM)](https://documentation.marine.copernicus.eu/PUM/CMEMS-MED-PUM-006-004.pdf).

Reviewer: Thanks! I will probably structure it a little differently in the vignette and I will include a link to this tutorial as it is a wonderful practical showcase. By the way, do you have a time-schedule for publishing these tutorials?

I'll see if I can spend some time on this in the Easter weekend before submitting it to CRAN.

#| fig-height: 3.5
ggplot(benthos_occurrences, aes(x = year)) +
  geom_bar(aes(fill = after_stat(count)), width = 0.8) +
  scale_fill_continuous(guide = "none", trans = "reverse") +
  scale_x_continuous(breaks = sort(unique(benthos_occurrences$year))) +
  labs(x = "Year", y = "Number of records") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

Sampling is concentrated in two periods: a large 1998 campaign and a main cluster from 2004 to 2018 with internal gaps. We download yearly temperature spanning 1998 to 2018, covering both periods.

### Downloading Bottom Temperature

`cms_download_subset()` returns a [`stars`](../concepts.qmd#stars---spatiotemporal-arrays) object (a multi-dimensional array). Note that the `timerange` dates must match the layer's time steps, which as we saw above are indexed to January 1st of each year:

```{r}
#| label: download-bottom-temperature
bottom_temp_stars <- cms_download_subset(
  product = "MEDSEA_MULTIYEAR_PHY_006_004",
  layer = "cmems_mod_med_phy-tem_my_4.2km_P1Y-m",
  variable = "bottomT",
  region = study_bbox
  # TODO: restore timerange when

>   # https://github.com/pepijn-devries/CopernicusMarine/issues/164 is fixed

Reviewer: I think it is fixed in the latest github version. @annakrystalli could you please confirm?

Author: Confirmed fixed! And reverted back to temporal filtering

  # timerange = c("1998-01-01", "2018-01-01")
)
```

### Inspecting the Stars Object

Let's examine the structure. Printing a `stars` object shows its attribute (the variable, with summary statistics) and its dimensions:

```{r}
#| label: inspect-stars-object
bottom_temp_stars
```

`st_dimensions()` shows the three dimensions of the object:

```{r}
#| label: inspect-stars-dims
stars::st_dimensions(bottom_temp_stars)
```

Each yearly time step is a 2D spatial grid (longitude × latitude), and the full object stacks these along the time dimension. We can inspect the time values directly:

```{r}
#| label: inspect-stars-time
time_vals <- stars::st_get_dimension_values(bottom_temp_stars, "time")
# Ensure consistent POSIXct format (stars may use different time

> # reference systems depending on the data source and platform)

Author: Thanks for the update. Unfortunately we're still getting `units` objects rather than POSIXct on CI (Ubuntu) with the latest dev version (0.4.5.4), so we're keeping the `as.POSIXct()` for now. I posted details on pepijn-devries/CopernicusMarine#164.

sampled_years <- unique(benthos_occurrences$year)
keep <- as.integer(format(time_vals, "%Y")) %in% sampled_years
bottom_temp_sampled <- bottom_temp_stars[,,, keep]

cat("Retained", sum(keep), "of", length(keep), "yearly time steps")
cat("Years kept:", sort(unique(as.integer(format(time_vals[keep], "%Y")))))
```

`stars` objects are indexed as `[attribute, dim1, dim2, dim3]`, where the first position selects attributes (variables like `bottomT`) and the rest select along the spatial and temporal dimensions. Leaving a position empty keeps all values along that axis. So `[,,,keep]` means: all attributes, all longitudes, all latitudes, filter time.

### Computing the Mean Temperature

`st_apply()` works like R's `apply()`: the first argument names the dimensions to **keep**, and the function is applied over all remaining dimensions. Here, `c("longitude", "latitude")` keeps the spatial dimensions and applies `mean` across time, collapsing all retained yearly time steps into a single 2D layer:

```{r}
#| label: compute-mean-temperature
bottom_temp_mean <- bottom_temp_sampled |>
  stars::st_apply(c("longitude", "latitude"), mean, na.rm = TRUE)
```

### Converting Stars to Terra Raster

> The rest of our analysis uses `terra` for raster operations (zonal statistics, extraction). To convert from `stars` to a `terra` `SpatRaster`, we need to address a grid compatibility issue: the Copernicus Marine grid is **rectilinear** (cell spacing along longitude and latitude is not perfectly uniform), but `terra::rast()` requires a **regular** grid with uniform cell sizes.

Reviewer: It needs to be clear here what your target CRS is. The bottomT data **IS** in a regular grid, but only in the lat-lon-plane. So I find this text a bit confusing.
Essentially, for your analyses, you need all raster data in the same raster. What the exact properties are of source raster data may just distract from your message. You have picked the target CRS for a reason, and you need to have all raster data in a consistent grid in that CRS. Since the bottomT data does not comply, you will have to warp it to the desired raster

Author: Thanks for the correction. Rewrote this section: the issue is that `stars` represents the grid as explicit cell intervals rather than a regular definition, so `terra::rast()` can't accept it directly. Also replaced the hardcoded `1/24` with dynamically derived cell sizes. The goal of this step is purely to regularise the grid, not to reproject; the target grid inherits its CRS from the source bbox, and the actual CRS harmonisation to EPSG:3035 happens later during zonal statistics.

> `st_warp()` resamples a `stars` object onto a target grid. We build that target with `st_as_stars()`, passing `st_bbox(bottom_temp_mean)` to preserve the original spatial extent and setting `dx` and `dy` to 1/24° to match the native Copernicus Marine resolution (~4.2 km at this latitude):

Author: Done, added package prefixes for code-link.

  st_make_valid() |>
  st_simplify(dTolerance = 200) |>
  st_transform(4326)
```

For the occurrence points, we bin continuous body sizes into discrete size classes using `cut()`, creating categories that are easy to distinguish on a colour scale, and reproject to WGS 84:

```{r}
#| label: prepare-occurrences-leaflet
benthos_leaflet <- benthos_joined |>
  filter(!is.na(max_body_size_mm)) |>
  mutate(
    size_class = cut(
      max_body_size_mm,
      breaks = c(0, 10, 50, 200, 1000, Inf),
      labels = c("<10mm", "10-50mm", "50-200mm", "200-1000mm", ">1000mm"),
      right = FALSE
    )
  ) |>
  st_transform(4326)
```

We build the map in layers. `addPolygons()` draws the habitat polygons, using `colorFactor()` to map EUNIS habitat codes to colours from the `"turbo"` palette. `addCircleMarkers()` overlays occurrence points coloured by body size class. Both layers include clickable pop-ups with relevant attributes:

```{r}
#| label: fig-interactive-map
#| eval: false

# Colour palettes
habitat_pal <- colorFactor("turbo", domain = habitats_leaflet$eunis2019c)
size_pal <- colorFactor("viridis", domain = benthos_leaflet$size_class)

leaflet() |>

>   addProviderTiles("Esri.WorldGrayCanvas") |>

Reviewer: ```suggestion
  addProviderTiles("Esri.WorldGrayCanvas") |>
    addCmsWMTSTiles(
    "MEDSEA_MULTIYEAR_PHY_006_004",
    "cmems_mod_med_phy-tem_my_4.2km_P1Y-m",
    "bottomT") |>
```
